[SPARK-6910] [SQL] Support for pushing predicates down to metastore for partition pruning

piaozhexiu opened the pull request:
This PR supersedes my old one #6921. Since my patch has changed quite a bit, I am opening a new PR to make it easier to review.

The changes include-
- Implement `toMetastoreFilter()` function in `HiveShim` that takes `Seq[Expression]` and converts them into a filter string for Hive metastore.
  - This functions matches all the `AttributeReference` + `BinaryComparisonOp` + `Integral/StringType` patterns in `Seq[Expression]` and fold them into a string.
- Change `hiveQlPartitions` field in `MetastoreRelation` to `getHiveQlPartitions()` function that takes a filter string parameter.
- Call `getHiveQlPartitions()` in `HiveTableScan` with a filter string.

But there are some cases in which predicate pushdown is disabled-

| Case | Predicate pushdown |
| --- | --- |
| Hive integral and string types | Yes |
| Hive varchar type | No |
| Hive 0.13 and newer | Yes |
| Hive 0.12 and older | No |
| convertMetastoreParquet=false | Yes |
| convertMetastoreParquet=true | No |

In case of `convertMetastoreParquet=true`, predicates are not pushed down because this conversion happens in an `Analyzer` rule (`HiveMetastoreCatalog.ParquetConversions`). At this point, `HiveTableScan` hasn't run, so predicates are not available. But reading the source code, I think it is intentional to convert the entire Hive table w/ all the partitions into `ParquetRelation` because then `ParquetRelation` can be cached and reused for any query against that table. Please correct me if I am wrong.

cc @marmbrus

Commit:
Predicate pushdown into Hive metastore

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36516/consoleFull) for   PR 7216 at commit [`dce7abf`](https://github.com/apache/spark/commit/dce7abf40003ad36675034e9467a5003f7e551ad).

SparkQA commented:
[Test build #36516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36516/console) for   PR 7216 at commit [`dce7abf`](https://github.com/apache/spark/commit/dce7abf40003ad36675034e9467a5003f7e551ad).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala:
```diff
@@ -99,6 +104,56 @@ private[hive] object HiveShim {
+    // So if Hive matastore version is older than 0.13, predicates cannot be pushed down.
+    // See HIVE-4888.
+    val versionPattern = "([\\d]+\\.[\\d]+).*".r
+    hiveMetastoreVersion match {
```
Instead of checking the version here, I'd see if we can't move this into the [per version HiveShim](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala) and just override it in the specific version classes, just so that we can be consistent with all of the other version specific code.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala:
```diff
@@ -99,6 +104,56 @@ private[hive] object HiveShim {
+            curr match {
+              case Some(currStr) if (prevStr.nonEmpty) => s"$prevStr and $currStr"
+              case Some(currStr) if (prevStr.isEmpty) => currStr
+              case None => prevStr
```
Is some of this logic an elaborate way to do `listOfStringPredicates.mkString(" and ")`?

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -689,6 +666,34 @@ private[hive] case class MetastoreRelation
     }
   )
 
+  def getHiveQlPartitions(filter: Option[String]): Seq[Partition] = {
```
minor: I'd consider having this function also take an `Expression` or a `Seq[Expression]` so that the conversion from expressions to string is isolated from the rest of the code.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientInterface.scala:
```diff
@@ -71,7 +71,12 @@ private[hive] case class HiveTable(
   def isPartitioned: Boolean = partitionColumns.nonEmpty
 
-  def getAllPartitions: Seq[HivePartition] = client.getAllPartitions(this)
+  def getPartitions(filter: Option[String]): Seq[HivePartition] = {
```
perhaps a default argument value of `None`?

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala:
```diff
@@ -316,6 +313,13 @@ private[hive] class ClientWrapper(
 
+  override def getPartitionsByFilter(
+      hTable: HiveTable,
+      filter: String): Seq[HivePartition] = withHiveState {
```
Even this filter could probably still be an `Expression`.

marmbrus commented:
This is looking very close!  Thanks again for doing such a thorough evaluation of the differing things Hive supports here. :)

Commit:
Incorporate review comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37074 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37074/consoleFull) for   PR 7216 at commit [`f134e91`](https://github.com/apache/spark/commit/f134e91b23a03d75ef5889c9f77ea5652e48e473).

Commit:
Fix ordering of imports

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37074 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37074/console) for   PR 7216 at commit [`f134e91`](https://github.com/apache/spark/commit/f134e91b23a03d75ef5889c9f77ea5652e48e473).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #37081 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37081/consoleFull) for   PR 7216 at commit [`28c8843`](https://github.com/apache/spark/commit/28c884365cd85f97a95977daf48478d0e93917b1).

SparkQA commented:
[Test build #37081 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37081/console) for   PR 7216 at commit [`28c8843`](https://github.com/apache/spark/commit/28c884365cd85f97a95977daf48478d0e93917b1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

piaozhexiu commented:
@marmbrus Thank you for reviewing my patch. I think I addressed all your comments. Can you please take a look again?

Btw, the last jenkins failures seems unrelated to my PR. In fact, jenkins build had succeeded first and then failed in the 2nd run triggered by my minor commit that fixed imports order.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37092/consoleFull) for   PR 7216 at commit [`9f5feec`](https://github.com/apache/spark/commit/9f5feec86a6f93e590b4c7ff0200632b2b1eb9c4).

SparkQA commented:
[Test build #37092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37092/console) for   PR 7216 at commit [`9f5feec`](https://github.com/apache/spark/commit/9f5feec86a6f93e590b4c7ff0200632b2b1eb9c4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

piaozhexiu commented:
Isn't this a test issue? I keep getting the following pyspark error. Looks like it is because of a minor difference in two datetimes ``532502 != 532501``. The test was added on Jul 9th by commit c9e2ef52bb54f35a904427389dc492d61f29b018.
```
FAIL: test_time_with_timezone (__main__.SQLTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/tests.py", line 718, in test_time_with_timezone
    self.assertEqual(now, now1)
AssertionError: datetime.datetime(2015, 7, 11, 12, 23, 36, 532502) != datetime.datetime(2015, 7, 11, 12, 23, 36, 532501)
```

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37100 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37100/consoleFull) for   PR 7216 at commit [`254d62a`](https://github.com/apache/spark/commit/254d62a2dfbe46358f7ed5899260018e6ad9672a).

SparkQA commented:
[Test build #37100 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37100/console) for   PR 7216 at commit [`254d62a`](https://github.com/apache/spark/commit/254d62a2dfbe46358f7ed5899260018e6ad9672a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1054 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1054/consoleFull) for   PR 7216 at commit [`254d62a`](https://github.com/apache/spark/commit/254d62a2dfbe46358f7ed5899260018e6ad9672a).

SparkQA commented:
[Test build #1054 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1054/console) for   PR 7216 at commit [`254d62a`](https://github.com/apache/spark/commit/254d62a2dfbe46358f7ed5899260018e6ad9672a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37133/consoleFull) for   PR 7216 at commit [`aa1490f`](https://github.com/apache/spark/commit/aa1490f796494bad889025026875a7b6fffff6b9).

SparkQA commented:
[Test build #37133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37133/console) for   PR 7216 at commit [`aa1490f`](https://github.com/apache/spark/commit/aa1490f796494bad889025026875a7b6fffff6b9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Least(children: Expression*) extends Expression `
  * `case class Greatest(children: Expression*) extends Expression `

AmplabJenkins commented:
Merged build finished. Test PASSed.

marmbrus commented:
Thanks!  I'm going to merge this to master.

rxin commented:
@piaozhexiu can you close this pr since github is not closing it automatically? Thanks!

piaozhexiu commented:
Sure. Thank you for merging it!
